pod/annotations.go

```diff
@@ -72,8 +72,11 @@ const (
 	AnnotationKeyJobType                = "v3.job.titus.netflix.com/type"
 	AnnotationKeyJobDescriptor          = "v3.job.titus.netflix.com/descriptor"
 	// AnnotationKeyPodTitusContainerInfo - to be removed once VK supports the full pod spec
-	AnnotationKeyPodTitusContainerInfo            = "pod.titus.netflix.com/container-info"
+	AnnotationKeyPodTitusContainerInfo = "pod.titus.netflix.com/container-info"
+	// AnnotationKeyPodTitusEntrypointShellSplitting tells the executor to preserve the legacy shell splitting behaviour
 	AnnotationKeyPodTitusEntrypointShellSplitting = "pod.titus.netflix.com/entrypoint-shell-splitting-enabled"
+	// AnnotationKeyPodTitusUserEnvVarsStartIndex tells the executor what index the user-specified environment variables start at
```

Comment:
Hmm. I guess this won't break as long as whatever is creating this annotation (the CP) is always doing:

```
env := append(TitusProvidedEnv, UserProvidedEnv...)
user-env-vars-index = len(UserEnv) 
```
?

I guess we just just have to be really careful about always making sure the CP does it that way.
And if we ever mutate a pod and add an env... how will that still validate?

Makes me think there should be a UserProvideEnv annotation (maybe base64ed) that, most of the time will be super small, but we can be 100% sure nobody is going to try to like, inject vars into such a thing, and it will *always* pass metatron validation.

Reply:
I'd really like the "pods" that we use in Titus to be as close as possible to real pods - in part so that if we move to RK one day, we can, and in part so that people who are familiar with k8s but not with Titus can think about it in the same way.

> And if we ever mutate a pod and add an env... how will that still validate?

Hmm, good point.  I guess we could do one of 2 things:

1) Ensure that if we're adding env vars via a webhook, that they get added in the "titus" section, and we update the index annotation.

2) Make the annotation the name of the first user env var.  We'd still need to make sure that env vars added by the webhook are spliced into the "titus" section, but at least we wouldn't have to update the index.  So this would look like:
```
pod.titus.netflix.com/user-env-vars-start-name: "MY_USER_ENV_VAR"
```
We would, however, need to make sure that if we're adding an env var in the webhook that has the same name as `MY_USER_ENV_VAR`, that we remove the user's `MY_USER_ENV_VAR`, and bump the start name to whatever the next env var is named.

I thought the second would be cleaner, but the duplicate env var thing would be pretty gross.  What do you think?

Reply:
> I'd really like the "pods" that we use in Titus to be as close as possible to real pods - in part so that if we move to RK one day, we can, and in part so that people who are familiar with k8s but not with Titus can think about it in the same way. Also, I'd like us to fight against having yet another base64'ed annotation. :)
> 

No no, I'm suggesting that we would have the normal env field that any human would understand, and a frozen annotation just for metatron.

```
spec:
  env:
    TITUS_THING=foo
    TITUS_RAM=bar
    ... 15 things
    USER_ENV=1
    SIDECAR_ENV=1
    ARCHER_ADDED_THIS=1
    WHO_KNOWS_WHAT_ELSE=1
    SOMEONE_PUT_THIS_ONE_CONTAINER_IN_DEBUG_MODE=true
  annotation:
    metatron.netflix.com/original_user_provided_env: base64(USER_ENV=1)
```

And when we go to check metatron, we only use the metatron annotation.

Reply:
What do we think about using the `EnvVarSource` [field](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#envvar-v1-core) and populating it with a dummy hint (user vs Titus-provided) possibly with a made up `ConfigMapKeyRef` that VK is not going to resolve.
Also in case of RK, the dummy ref might still be unused as the `EnvVar.value` will always be populated as per the documentation.
> `valueFromEnvVarSource` | Source for the environment variable's value. Cannot be used if value is not empty.
I haven't tried it but might work consideration?

Reply:
The goals of this change in contract IMO are:
* keep the contract as close to the pod spec
* keep the pod object as small as possible

With those goals in mind, it makes sense to add an annotation to keep metatron working but with a mechanism that is small. I don't want to go down the path of encoding any more large user fields as that what we are trying to get away from. However if the index convention is too much of a convention, we could add an annotation that containers a delimited set of keys that are system env vars:
`pod.titus.netflix.com/system-env-vars-names: "key1,key2,key3"`

I like the index approach though as it also means that env vars should have ordering which I think makes it easier for me to debug.

Reply:
@joshi-keyur, `EnvVarSource` is a good idea, but I think it violates the principle of keeping the contract as close as possible to the "real" k8s pod spec.

We can always start with the index method and if it turns out to be super painful, move to another solution in the next version of the pod spec.